rfc-116-store-attachment-data-in-content-items.md

```diff
@@ -0,0 +1,321 @@
+
+
+## Problem
+
+Documents on GOV.UK can have attachments, which come in a few
+different types.  Whitehall, which has the richest model of
+attachments, has:
+
+- External attachments (links to other websites)
+- HTML attachments
+- File attachments (which can have previews)
+
+When thinking about how attachments are displayed to the user, there
+are two sorts:
+
+- *Document-level attachments*, such on Whitehall publication and
+  consultation documents, which are displayed separately to the main
+  content.
+
+- *Inline attachments*, such as in other Whitehall documents, manual
+  sections, specialist documents, and travel advice documents, which
+  are referenced in the page content and rendered as links.
+
+Publishing applications vary in how easily accessible they make
+attachment metadata in their content items.  Whitehall only sends
+rendered HTML to the Publishing API.  Other publishing apps send a
+combination of rendered HTML and metadata.
+
+This inconistency restricts what we can do with attachments,
+particularly for Whitehall content.  For example, we cannot generate
+comprehensive [schema.org][] metadata for attachments: [see this
+comment][].
+
+Additionally, users of the content API cannot make use of attachments
+without parsing the document body.  This inhibits creative use of our
+content.
+
+We should make metadata for Whitehall attachments available in the
+same way as other publishing applications, but that requires some
+changes into how we represent attachments across the stack.
+
+
+## Proposal
+
+In our content schemas we already have an [`asset_link` type][], which
+is used for attachments in specialist documents and manual sections.
+It's also used in travel advice pages for the page image and a single
+attachment.  This type appears to solve all of our problems:
+[Publishing API can render such attachments with govspeak][], and
+content items can expose this data to enable programmatic use.
+
+However, the metadata Govspeak expects and the `asset_link` type are
+only *informally* the same.  And furthermore, Whitehall has much richer
+attachment metadata than either `asset_link` or Govspeak allow.
+
+Given that `asset_link` is currently used for both images and
+attachments, adding all of the missing metadata to `asset_link` feels
+a mistake: images and attachments have different metadata, so we
+shouldn't conflate the two.
+
+I propose we do this:
+
+1. Split `asset_link` into four new types: `base_attachment_asset`,
+   `attachment_asset`, `specialist_publisher_attachment_asset` and
+   `image_asset`, updating the schemas which use `asset_link`
+   accordingly, with these fields:
+
+   - `base_attachment_asset`:
+     - `url` (required)
+     - `content_type` (required)
+     - `title`
+   - `attachment_asset` (extends `base_attachment_asset`):
+     - no extra fields at this time
+   - `specialist_publisher_attachment_asset` (extends `base_attachment_asset`):
```

Comment:
Can we add some indicator about the need for this type? It's the first time we've mentioned Specialist Publisher in the RFC, and I think future readers would find it helpful to understand the additional persistence problem I believe we're trying to solve here.

Reply:
:+1: